### DISCUSSION TITLE
How do you handle MultiPartitioned Assets keys when executing as Single-Run vs Multi-Run?

### DISCUSSION THREAD

**Discussion by JoeNaso** [dagster-io/dagster#17517] [Q&A] [answered]
When using a Multipartition definition like below, single runs fail when accessing the partition values as listed in the docs. 

```
rollup_partition = MultiPartitionsDefinition(
    {
        "target_date": DailyPartitionsDefinition(start_date="2023-10-25"),
        "origin": StaticPartitionsDefinition(["some", "static", "values"]),
    }
)
@asset(partitions_def=rollup_partition)
def do_work(context):
    #this will fail on a single run but works for multi-run
   partitions = context.partition_key.keys_by_dimension
```
What's the recommended way to handle multi-partitioned asset keys when we want to be able to run this job ad-hoc as a single run or a multi-run? 

I realize there are `asset_partition_key_range_for_output()` and `asset_partitions_time_window_for_output()` context methods, but they can be unwieldy. I would expect the multi-partition key API to be the same regardless of the run "type".

**Comment by sryza** [comment, marked as answer]
Hey Joe, in a single-run backfill, the `@asset`-decorated function is called just once, rather than once per partition. So if the backfill covers multiple partitions, then that invocation of the function is expected to process all the partitions at once.  In that case, there's not 
a just single partition key associated with the invocation, so you need to use `context.partition_key_range` or `context.partition_keys`.
